pulp_rpm/app/models/distribution.py

```diff
@@ -328,3 +331,21 @@ class Meta:
             "packages",
             "distribution_tree",
         )
+
+
+@receiver(pre_delete, sender=DistributionTree)
+def cleanup_subrepos(sender, instance, **kwargs):
+    """
+    Remove subrepos when a DistributionTree is being removed.
+
+    """
+    with transaction.atomic():
+        for variant in instance.variants.all():
+            if variant.repository:
+                variant.repository.delete()
+            variant.delete()
+        for addon in instance.addons.all():
+            if addon.repository:
+                addon.repository.delete()
+            addon.delete()
+        # remove content similar to orphan cleanup
```

Comment:
When subrepos are removed, now there is more orphaned content.
Ask users to run a cleanup again?
Remove content here similar to orphan cleanup? Is it safe? can there be a race condition? e.g. I remove a distribution tree and content related to subrepos and at the same time other repository sync just checked that a package (from a subrepo) already exists in Pulp.

Reply:
What bout if the orphan cleanup code itself detected an orphan it's about to delete is a DistributionTree, and if so perform the orphan cleanup again until no DistributionTree units  are to be cleaned up?

Reply:
I'm not against that. Orphan cleanup code is in the pulpocore.
Do you suggest to incorporate into the pulpcore orphan cleanup a plugin input based on which orphan clean up will try to clean more content?

Reply:
hmmmmmmm  (that's all I have right now) :smile: